### PR TITLE
New Spezi Standard Implementation

### PR DESCRIPTION
**Pull request by niallkehoe**
# *New Spezi Standard Implementation*

## :recycle: Current situation & Problem
*Currently, the SpeziQuestionnaire module lacks support for the new Spezi Standard released in Spezi 0.7.0*
*This PR implements the improvements mentioned in [https://github.com/StanfordSpezi/Spezi/pull/73](Pull 73) *

## :bulb: Proposed solution
*Update the codebase to the newest Spezi 0.7.0*

## :gear: Release Notes 
This feature will allow for greater flexibility when using Spezi Questionnaire.

```
/// A Standard which all Spezi Questionnaires must follow
public protocol QuestionnaireConstraint: Standard {
    
    /// Adds a new `QuestionnaireResponse` to the ``QuestionnaireDataSource``
    /// - Parameter response: The `QuestionnaireResponse` that should be added.
    func add(_ response: QuestionnaireResponse) async
}
```

### Related PRs
*See [https://github.com/StanfordSpezi/Spezi/pull/73](Pull 73) in Spezi for details*

### Testing
*Passes all tests*

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [ ✅ ] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PSchmiedmayer** [comment]
@PSchmiedmayer Take a look at the test coverage

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@niallkehoe Thank you for improving Spezi and bringing Spezi Questionnaire up to Spezi 0.7.0! Thank you for removing the constraint and simplifying the questionnaire module!

I have taken a look at the failing UI tests and it seems like the main reason was the SwiftLint task in the Xcode project that just lined all downloaded dependencies. I have added the right ignores in the SwiftLint files. I have also explicitly added the package target to the list of build steps to ensure that the code coverage is correctly reported.

I had a few comment rating from style-level comments, documentation, to the UI tests to verify the functionality. Once those comments are addressed we can merge the PR and tag a new release. Feel free to reach out if you have any questions or comments regarding the feedback, always happy to help and support!

Looking forward to your first Spezi contribution, more to come! 🚀

File: Sources/SpeziQuestionnaire/QuestionnaireDataSource.swift [hunk 1]
```diff
@@ -8,14 +8,13 @@
...
 import Spezi
-import SpeziFHIR
 import SwiftUI
 
 
 /// Maps `Questionnaires` returned from the ``QuestionnaireView`` to the `Standard` of the Spezi application.
 ///
 /// Use the ``QuestionnaireDataSource/init(adapter:)`` initializer to add the data source to your `Configuration`.
-/// You can use the ``QuestionnaireDataSource/init()`` initializer of you use the FHIR standard in your Spezi application:
+/// You can use the ``QuestionnaireDataSource/init()`` initializer of you use the Questionnaire standard in your Spezi application:
```
[the remark this answers is not in the document]
Comment (niallkehoe) [reply Sources/SpeziQuestionnaire/QuestionnaireDataSource.swift:17]: Not sure what you had in mind. Would "QuestionnaireDataSource uses QuestionnaireConstraint which can be customized by conforming your standard to QuestionnaireConstraint"?

File: Tests/UITests/TestApp/TestAppDelegate.swift [hunk 2]
```diff
@@ -7,14 +7,17 @@
...
 
 import Spezi
-import SpeziFHIR
 import SpeziQuestionnaire
 import SwiftUI
 
+/// an example Standard used for the configuration
+actor ExampleStandard: Standard {
+    // ...
+}
```
[the remark this answers is not in the document]
Comment (niallkehoe) [reply Tests/UITests/TestApp/TestAppDelegate.swift]: I tried including a counter which increases on add(), however in this branch and main the template application never finishes the questionnaire. After the last question, there is a breakpoint on QuestionnaireView 78: questionnaireDataSource.add(questionnaireResponse): 
2023-07-31 11:14:57.820054-0700 TestApp[42944:2475753] SwiftUI/EnvironmentObject.swift:82: Fatal error: No ObservableObject of type QuestionnaireDataSource found. A View.environmentObject(_:) for QuestionnaireDataSource may be missing as an ancestor of this view.

**Comment by vishnuravi** [comment]
Thank you @niallkehoe!

File: Sources/SpeziQuestionnaire/QuestionnaireConstraint.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import ModelsR4
+import Spezi
+
+
+/// A Constraint which all `Standard` instances must conform to when using the Spezi Questionnaire module.
+public protocol QuestionnaireConstraint: Standard {
+    /// Adds a new `QuestionnaireResponse` to the `Standard` conforming to `QuestionnaireConstraint`.
+    /// - Parameter response: The `QuestionnaireResponse` that should be added.
+    func add(_ response: QuestionnaireResponse) async
```
Comment (PSchmiedmayer) [inline Sources/SpeziQuestionnaire/QuestionnaireConstraint.swift:17]: In sync to the HealthKit PR, I would suggest that we slightly adjust the function signature here to make it easier to multiplex between the different add methods we might need to implement in a `Standard` conforming to the different constraints:
```suggestion
    func add(response: QuestionnaireResponse) async
```

File: Tests/UITests/TestApp/ContentView.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    @State var displayQuestionnaire = false
+    
+    
+    var body: some View {
+        Text("No. of surveys complete: \(surveyResponseCount)")
+            .onReceive(standard.objectWillChange) { _ in
+                Task { @MainActor in
+                    surveyResponseCount = await standard.surveyResponseCount
+                }
+            }
```
Comment (PSchmiedmayer) [inline Tests/UITests/TestApp/ContentView.swift:26]: Now that you have enforced that `surveyResponseCount` can only be accessed from the `@MainActor`, Swift is actually smart enough to see that this closure will also always be executed in the main actor (enforced by SwiftUI) and you can therefore access the property without the need for an `await`:
```suggestion
        Text("No. of surveys complete: \(standard.surveyResponseCount)")
```

File: Tests/UITests/TestApp/ContentView.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+//
+
+import FHIRQuestionnaires
+import SpeziQuestionnaire
+import SwiftUI
+
+
+struct ContentView: View {
+    @EnvironmentObject var standard: ExampleStandard
+    @State var surveyResponseCount = 0
```
Comment (PSchmiedmayer) [inline Tests/UITests/TestApp/ContentView.swift:16]: Once you made the changes to the `Text`, you can remove this property there 🚀

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for all the changes and great additions @niallkehoe! I only had two very minor change requests that make the code a bit cleaner and ensure that you can reduce some of the complexity in the test views.
Once those are addressed you can resolve the conversations and merge the PR. Thank you for the contribution and taking the time to improve Spezi! 🚀

**Pull request merged**